pkg/eventprocessor/impl/executor/impl/executor.go

```diff
@@ -80,6 +86,8 @@ func NewExecutor(
 func (ex *Executor) NewBlockScope(ctx context.Context, newBlockNum int64) (executor.BlockScope, error) {
 	select {
 	case <-ex.chBlockScope:
+	case <-ex.closed:
+		return nil, fmt.Errorf("executor is closed")
```

Comment:
If `ex.closed` is closed, then we simply return an error instead of panicking.

The panicking makes sense to detect if other client is trying to call `NewBlockScope` concurrently, which is definitely wrong and a bug.

But notice how `Close()` works. It `<-ex.chBlockScope` to be sure there's no ongoing execution of a block, so it closes when no block is being executed. But... that means that a concurrent `NewBlockScope` could also happen at the same time, making `case <-ex.chBlockScope:` in L88 fail, thus defaulting to panic which isn't entirely correct. (There's no serial execution, but just being unlucky that we're closing).

This was fixed by an extra `ex.close` channel that is first closed by `Closed()`, and then it does L156. This way, before doing L156, we know that any concurrent `NewBlockScope` will error (instead of panicking) which is way better for graceful closing.

LMK if this sounds confusing!